packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js

```diff
@@ -0,0 +1,213 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useSelect } from '@wordpress/data';
+import { useState } from '@wordpress/element';
+import {
+	Tooltip,
+	__experimentalNavigationGroup as NavigationGroup,
+	__experimentalNavigationItem as NavigationItem,
+} from '@wordpress/components';
+import { Icon, home, plus, undo } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import TemplatePreview from './template-preview';
+import ThemePreview from './theme-preview';
+
+const TEMPLATE_OVERRIDES = {
+	page: ( slug ) => `page-${ slug }`,
+	category: ( slug ) => `category-${ slug }`,
+	post: ( slug ) => `single-post-${ slug }`,
+};
+
+function TemplateLabel( { template, homeId } ) {
+	return (
+		<>
+			{ template.slug }{ ' ' }
+			{ template.id === homeId && (
+				<Tooltip text={ __( 'Home' ) }>
+					<div className="edit-site-template-switcher__label-home-icon">
+						<Icon icon={ home } />
+					</div>
+				</Tooltip>
+			) }
+			{ template.status !== 'auto-draft' && (
+				<Tooltip text={ __( 'Customized' ) }>
+					<span className="edit-site-template-switcher__label-customized-dot" />
+				</Tooltip>
+			) }
+		</>
+	);
+}
+
+function NavigationItemWithIcon( { icon, title, ...props } ) {
+	return (
+		<NavigationItem
+			title={
+				<>
+					{ title }{ ' ' }
+					{ icon && (
+						<div className="edit-site-template-switcher__label-home-icon">
+							<Icon icon={ icon } />
+						</div>
+					) }
+				</>
+			}
+			{ ...props }
+		/>
+	);
+}
+
+export default function TemplateSwitcher( {
+	page,
+	activeId,
+	onActiveIdChange,
+	onActiveTemplatePartIdChange,
+	onAddTemplate,
+	onRemoveTemplate,
+} ) {
+	const [ hoveredTemplate, setHoveredTemplate ] = useState();
+	const [ themePreviewVisible, setThemePreviewVisible ] = useState( false );
+
+	const onHoverTemplatePart = ( id ) => {
+		setHoveredTemplate( { id, type: 'template-part' } );
+	};
+	const onMouseEnterTheme = () => {
+		setThemePreviewVisible( () => true );
+	};
+	const onMouseLeaveTheme = () => {
+		setThemePreviewVisible( () => false );
+	};
```

Comment:
What about refactoring this a little bit?

Now we don't have a `onHover` anymore, so we should probably define both on enter/on leave for the template part as well.
Also we could simplify `setHoveredTemplate` since it's only ever used for template parts.
Finally, I'm not sure why `setThemePreviewVisible` uses a callback, when it doesn't derive the new value from the previous. 🤔 

E.g.
```jsx
const [ templatePartPreviewVisible, setTemplatePartPreviewVisible ] = useState();
const [ themePreviewVisible, setThemePreviewVisible ] = useState( false );

const onMouseEnterTemplatePart = ( id ) => setTemplatePartPreviewVisible( id );
const onMouseLeaveTemplatePart = ( id ) => setTemplatePartPreviewVisible( null );

const onMouseEnterTheme = ( id ) => themePreviewVisible( true );
const onMouseLeaveTheme = ( id ) => themePreviewVisible( false );
```